odl/discr/partition.py

```diff
@@ -0,0 +1,561 @@
+
+        >>> part = RectPartition(grid, begin=[0, -1], end=[1, 2])
+        >>> part.cell_boundaries()
+        (array([ 0. ,  0.5,  1. ]), array([-1. , -0.5,  1. ,  2. ]))
+        """
+        bdry_vecs = []
+        for ax, vec in enumerate(self.grid.coord_vectors):
+            bdry = np.empty(len(vec) + 1)
+            bdry[1:-1] = (vec[1:] + vec[:-1]) / 2.0
+            bdry[0] = self.min()[ax]
+            bdry[-1] = self.max()[ax]
+            bdry_vecs.append(bdry)
+
+        return tuple(bdry_vecs)
+
+    def cell_sizes(self):
+        """Return the cell sizes as coordinate vectors.
+
+        Returns
+        -------
+        csizes : `tuple` of `numpy.ndarray`
+            The cell sizes per axis. The length of the vectors is the
+            same as the corresponding ``grid.coord_vectors``.
+            For axes with 1 grid point, cell size is set to 0.0.
+
+        Examples
+        --------
+        >>> grid = TensorGrid([0, 1], [-1, 0, 2])
+        >>> part = RectPartition(grid)  # Implicit begin and end
+        >>> part.cell_sizes()
+        (array([ 1.,  1.]), array([ 1. ,  1.5,  2. ]))
+
+        >>> part = RectPartition(grid, begin=[0, -1], end=[1, 2])
+        >>> part.cell_sizes()
+        (array([ 0.5,  0.5]), array([ 0.5,  1.5,  1. ]))
+        """
+        csizes = []
+        for ax, vec in enumerate(self.grid.coord_vectors):
+            if len(vec) == 1:
+                csizes.append(np.array([0.0]))
+            else:
+                csize = np.empty_like(vec)
+                csize[1:-1] = (vec[2:] - vec[:-2]) / 2.0
+                csize[0] = (vec[0] + vec[1]) / 2 - self.min()[ax]
+                csize[-1] = self.max()[ax] - (vec[-2] + vec[-1]) / 2
+                csizes.append(csize)
+
+        return tuple(csizes)
+
+    @property
+    def cell_size(self):
+        """Size of the 'inner' cells, regardless of begin and end.
+
+        Only defined if ``self.grid`` is a `RegularGrid`.
+
+        Examples
+        --------
+        >>> grid = RegularGrid([0, 0], [1, 1], (3, 5))
+        >>> part = RectPartition(grid)
+        >>> part.cell_size
+        array([ 0.5 ,  0.25])
+        """
+
+        if not self.is_regular:
+            raise NotImplementedError(
+                'cell size not defined for irregular partitions. Use '
+                'cell_sizes() instead.')
+
+        on_boundary = self.nodes_on_boundary()
+        num_cells = []
+        for n, on_bdry in zip(self.shape, on_boundary):
+            num_on_bdry = sum(on for on in on_bdry)
+            num_cells.append(n - num_on_bdry / 2)
+
+        return self.extent() / num_cells
+
+    @property
+    def cell_volume(self):
+        """Volume of the 'inner' cells, regardless of begin and end.
+
+        Only defined if ``self.grid`` is a `RegularGrid`.
+
+        Examples
+        --------
+        >>> grid = RegularGrid([0, 0], [1, 1], (3, 5))
+        >>> part = RectPartition(grid)
+        >>> part.cell_volume
+        0.125
```

Comment:
This is true to some extent, but it's the same as the old grid implementation would have given.

It rather makes the choice of initializing partitions from grids only somewhat debatable. Perhaps it's better to add an `IntervalProd` argument, but in any case, to get an example where you partition exactly the rectangle [0, 1] x [0, 1] with equally sized cells requires calculating the corresponding grid points.